setup/mscs/msc3861/msc3861_user_verifier.go

```diff
@@ -0,0 +1,435 @@
+package msc3861
+
+import (
+	"cmp"
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"slices"
+	"strings"
+
+	"github.com/element-hq/dendrite/clientapi/auth"
+	"github.com/element-hq/dendrite/setup/config"
+	"github.com/element-hq/dendrite/userapi/api"
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/matrix-org/util"
+)
+
+const externalAuthProvider string = "oauth-delegated"
+
+// Scopes as defined by MSC2967
+// https://github.com/matrix-org/matrix-spec-proposals/pull/2967
+const (
+	scopeMatrixAPI          string = "urn:matrix:org.matrix.msc2967.client:api:*"
+	scopeMatrixGuest        string = "urn:matrix:org.matrix.msc2967.client:api:guest"
+	scopeMatrixDevicePrefix string = "urn:matrix:org.matrix.msc2967.client:device:"
+)
+
+type errCode string
+
+const (
+	codeIntrospectionNot2xx        errCode = "introspectionIsNot2xx"
+	codeInvalidClientToken         errCode = "invalidClientToken"
+	codeAuthError                  errCode = "authError"
+	codeMxidError                  errCode = "mxidError"
+	codeOpenidConfigEndpointNon2xx errCode = "openidConfigEndpointNon2xx"
+	codeOpenidConfigDecodingFailed errCode = "openidConfigDecodingFailed"
+)
+
+// MSC3861UserVerifier implements UserVerifier interface
+type MSC3861UserVerifier struct {
+	userAPI      api.UserInternalAPI
+	serverName   spec.ServerName
+	cfg          *config.MSC3861
+	httpClient   *http.Client
+	openIdConfig *OpenIDConfiguration
+	allowGuest   bool
+}
+
+func newMSC3861UserVerifier(
+	userAPI api.UserInternalAPI,
+	serverName spec.ServerName,
+	cfg *config.MSC3861,
+	allowGuest bool,
+	httpClient *http.Client,
```

Comment:
`httpClient` should probably be a `fclient.Client`, see https://github.com/matrix-org/gomatrixserverlib/blob/0a1b2bafb5cf72727787359d39653493b3318398/fclient/client.go#L44-L50

And it should be an error if it is not set, we shouldn't use `http.DefaultClient`.